posydon/popsyn/GRB.py

```diff
@@ -0,0 +1,119 @@
+
+def get_GRB_properties(df, GRB_efficiency, GRB_beaming, E_GRB_iso_min=0.):
+    """Compute GRB properties for a given binary population."""
+    
+    # check if radiated disk mass is avaialble to compute GRB properties
+    if ('S1_m_disk_radiated' not in df or
+        'S2_m_disk_radiated' not in df):
+        raise ValueError('m_disk_radiated not found in the dataframe!')
+    
+    # reminder the user about the threshold cut
+    if E_GRB_iso_min != 0.:
+        warnings.warn("We only consider GRBs with isotropic equivalent "
+                        f"energy larger than {E_GRB_iso_min} erg!")
+        
+    # define some methods
+    
+    def compute_eta(i, sel):
+        """Compute GRB energy conversion efficiency factor."""
+        if (not isinstance(GRB_efficiency, float) and
+            GRB_efficiency >= 0. and GRB_efficiency <= 1.):
+            raise ValueError(f'GRB efficiency {GRB_efficiency} '
+                             'must be float value in [0,1] range!')
+        # preset all efficiency factors to NaN and set them
+        # only for the selected binaries emitting a GRB
+        df[f'S{i}_eta'] = [np.nan] * df.shape[0]
+        df.loc[sel,f'S{i}_eta'] = GRB_efficiency
+    
+    def compute_beaming_factor(i, sel):
+        """Compute GRB beaming factor."""
+        # TODO: check out Github copilot suggestions: 
+        # 'Fong+15', 'Ghirlanda+16', 'Lamb+19'?
+        
+        df[f'S{i}_f_beaming'] = [np.nan] * df.shape[0]
+        if isinstance(GRB_beaming, float):
+            df.loc[sel, f'S{i}_f_beaming'] = GRB_beaming
+        elif GRB_beaming == 'Goldstein+15':
+            n = df.loc[sel,f'S{i}_f_beaming'].shape[0]
+            # generate a opening angle from a lognormal distribution 
+            # as in Table 3 of https://arxiv.org/pdf/1512.04464.pdf
+            theta = 10**(np.random.normal(0.77,0.37,n))/180.*np.pi # radian
+            # compute the solid angle, factor 2 to accounts for the two emispheres
+            solid_angle = 4*np.pi*(1.-np.cos(theta)) 
+            df.loc[sel,f'S{i}_f_beaming'] = solid_angle/(4*np.pi)
+        elif GRB_beaming == 'Lloyd-Ronning+19':
+            if z is None:
+                raise ValueError('Redshift not specified!')
+            else:
+                z = df.loc[sel,'z_formation']
+            n = df.loc[sel,f'S{i}_f_beaming'].shape[0]
+            # generate a opening angle from a lognormal distribution
+            # as in Table 3 of https://arxiv.org/pdf/1512.04464.pdf
+            theta =10**(np.random.normal(0.77*(1+z)**(-0.75),0.37,n))/180.*np.pi # radian
+            # compute the solid angle, factor 2 to accounts for the two emispheres
+            solid_angle = 4*np.pi*(1.-np.cos(theta))
+            df.loc[sel,f'S{i}_f_beaming'] = solid_angle/(4*np.pi)
+        elif GRB_beaming == 'Pescalli+15':
+            def f_B(E,eta):
+                xi=0.5
+                t=25.
+                C = (7e46)**xi
+                fb = np.ones(len(E))
```

Comment:
Please double check, that you like to have a beaming fraction of 1, in case of a negative energy. I'd more expect a fraction of 0 to cut off negative energies.